exp/addons/internal/controllers/clusterresourceset_scope.go

```diff
@@ -0,0 +1,164 @@
+}
+
+func newResourceReconcileScope(
+	clusterResourceSet *addonsv1.ClusterResourceSet,
+	resourceRef addonsv1.ResourceRef,
+	resourceSetBinding *addonsv1.ResourceSetBinding,
+	normalizedData [][]byte,
+	objs []unstructured.Unstructured,
+) resourceReconcileScope {
+	base := baseResourceReconcileScope{
+		clusterResourceSet: clusterResourceSet,
+		resourceRef:        resourceRef,
+		resourceSetBinding: resourceSetBinding,
+		data:               normalizedData,
+		normalizedObjs:     objs,
+		computedHash:       computeHash(normalizedData),
+	}
+	if clusterResourceSet.IsStrategy(addonsv1.ClusterResourceSetStrategyApplyOnce) {
+		return &reconcileApplyOnceScope{base}
+	} else if clusterResourceSet.IsStrategy(addonsv1.ClusterResourceSetStrategyReconcile) {
+		return &reconcileStrategyScope{base}
+	}
+
+	return nil
+}
+
+type baseResourceReconcileScope struct {
+	clusterResourceSet *addonsv1.ClusterResourceSet
+	resourceRef        addonsv1.ResourceRef
+	resourceSetBinding *addonsv1.ResourceSetBinding
+	normalizedObjs     []unstructured.Unstructured
+	data               [][]byte
+	computedHash       string
+}
+
+func (b baseResourceReconcileScope) objs() []unstructured.Unstructured {
+	return b.normalizedObjs
+}
+
+func (b baseResourceReconcileScope) hash() string {
+	return b.computedHash
+}
+
+type reconcileStrategyScope struct {
+	baseResourceReconcileScope
+}
+
+func (r *reconcileStrategyScope) needsApply() bool {
+	resourceBinding := r.resourceSetBinding.GetResourceBinding(r.resourceRef)
+
+	return resourceBinding == nil || !resourceBinding.Applied || resourceBinding.Hash != r.computedHash
+}
+
+func (r *reconcileStrategyScope) apply(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+	currentObj := &unstructured.Unstructured{}
+	currentObj.SetAPIVersion(obj.GetAPIVersion())
+	currentObj.SetKind(obj.GetKind())
+	err := c.Get(ctx, client.ObjectKeyFromObject(obj), currentObj)
+	if apierrors.IsNotFound(err) {
+		return createUnstructured(ctx, c, obj)
+	}
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"reading object %s %s/%s",
+			obj.GroupVersionKind(),
+			obj.GetNamespace(),
+			obj.GetName(),
+		)
+	}
+
+	patch := client.MergeFrom(currentObj.DeepCopy())
+	obj.SetResourceVersion(currentObj.GetResourceVersion())
+	if err = c.Patch(ctx, obj, patch); err != nil {
+		return errors.Wrapf(
+			err,
+			"patching object %s %s/%s",
+			obj.GroupVersionKind(),
+			obj.GetNamespace(),
+			obj.GetName(),
+		)
+	}
+
+	return nil
+}
+
+type reconcileApplyOnceScope struct {
+	baseResourceReconcileScope
+}
+
+func (r *reconcileApplyOnceScope) needsApply() bool {
+	return !r.resourceSetBinding.IsApplied(r.resourceRef)
+}
+
+func (r *reconcileApplyOnceScope) apply(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+	// The create call is idempotent, so if the object already exists
+	// then do not consider it to be an error.
```

Comment:
This doesn't seem the case anymore since we are dropping if !apierrors.IsAlreadyExists(err)  from createUnstructured; this makes the code, not re-entrant in case of errors.

Reply:
I think I'm missing something. I just moved `!apierrors.IsAlreadyExists(err)` up the chain one method to here. That keeps `createUnstructured` agnostic from the strategy and lets the scope decide what to do with the error. In this case, the apply once implementation will detect that the error is an "AlreadyExist" error (to be precised, it detects the returned error wraps an "AlreadyExist" error) and return nil. I added a unit test for this case.

Reply:
I'm not sure I follow where up in the chain is, because from what I can see
reconcileApplyOnceScope.applyObj is called by reconcileApplyOnceScope.apply (which invokes apply).
At this level, we already applied all the objects, but there is no handling of apierrors.IsAlreadyExists(err)
🤔

Reply:
note. idempotency is important to make this controller robust, so what about adding a test ensuring that the controller can recover from a failure in the middle of applying resources (it is most important for applying once and kind of an implicit behavior for reconcile, so might be it is already covered for the latter)

Reply:
Yup Fabrizio is right.

There are two usages of `createUnstructured`. l.127 and l.168. In l.168 we check for already exists in l.127 we don't.

Reply:
Ah now I got it. Not checking in l.127 is intentional as we want to surface the error in the ApplyOnce mode?

Reply:
To be honest I still wonder if implementing the "Reconcile" mode with SSA would be easier. We would just read the YAML and then send it with SSA and then SSA will take care of it.

SSA is GA since a while, it's becoming slowly the standard and at some point we might have to ask ourselves if we should migrate CRS over to SSA if we don't do it now already.

Reply:
Okay. I think this comes down to what should the expected behavior be if objects already exist.

My take:
* ApplyOnce
  * If the object already exists => update it
    * I think we should do that as it's called "ApplyOnce" not "CreateOnce"
  * else => create it
* Reconcile
  * If the object already exists => update it
  * else => create it

So I wonder if the logic in applyObj should be the same, with the only difference that based on `needsApply` we will only ever run the func once for `ApplyOnce`.

I realize that this would be a change to the `ApplyOnce` strategy, but to be honest I consider this a bugfix.

Reply:
cc @fabriziopandini @vincepri

Reply:
Regarding idempotency/re-entry, I understand your concerns and agree it's important to verify the algorithm correctness. In order to help with this, I added two tests (https://github.com/kubernetes-sigs/cluster-api/pull/7497/commits/9f84d7e6bbac70316cc87debeb2918705d4fcd8d):
* `ApplyOnce` CRS that tries to create a resource that already exists. It verifies that the other resources get created correctly and that the existing resource is not modify. This latter point follows the existing behavior, which if I remember correctly, it's by design as indicated in the original proposal.
* `ApplyOnce` CRS that tries to create a resource that fails. In order to do that, I specify such resource in a namespace that doesn't exist. I then verify that the other defined resources get created correctly and that the binding and the CRS condition reflect this error correctly. Then, the test creates the missing namespace and expect the resource to be created successfully.

I could create similar tests for the `Reconcile` mode, however, giving the nature of the continuous reconciliation not sure if it provides much value. There is already a test that verifies that the defined resources get updated when their definition gets updated when using `Reconcile`.

Btw this in parallel to @sbueringer last comment and just addressing the prior comments.

Reply:
@g-gaston Talked to Fabrizio about https://github.com/kubernetes-sigs/cluster-api/pull/7497?notification_referrer_id=NT_kwDOAEckWLI0NzU5MjE3NTU5OjQ2NjIzNjA#discussion_r1087716355

Turns out the current logic of ApplyOnce is intended and should stay that way.

The idea is that the controller should be re-entrant. Because of that we want to ignore already exist errors.

So let's just keep it as is

Reply:
Regarding: https://github.com/kubernetes-sigs/cluster-api/pull/7497#discussion_r1087745046

Looks good to me

Reply:
Regarding: https://github.com/kubernetes-sigs/cluster-api/pull/7497#discussion_r1087745046

Looks good to me